### PR TITLE
[FLOC-1153] Move package linting to during build.

### PR DESCRIPTION
**Pull request by tomprince**
https://clusterhq.atlassian.net/browse/FLOC-1153

### CODE REVIEW

File: admin/packaging.py [hunk 1]
```diff
@@ -478,6 +523,146 @@ def release(self):
...
 
+IGNORED_WARNINGS = {
+    PackageTypes.RPM: (
+        # Ignore the summary line rpmlint prints.
+        # We always check a single package, so we can hardcode the numbers.
+        '1 packages and 0 specfiles checked;',
+
+        # This isn't an distribution package so we deliberately install in /opt
+        'dir-or-file-in-opt',
+        # We don't care enought to fix this
```
Comment (wallrj) [inline admin/packaging.py]: typo "enought"

File: admin/packaging.py [hunk 2]
```diff
@@ -478,6 +523,146 @@ def release(self):
...
+        'no-copyright-file',
+
+        # These are in our dependencies.
+        'script-not-executable',
+        'embedded-javascript-library',
+        'extra-license-file',
+        'unstripped-binary-or-object',
+
+        # Werkzeug installs various images with executable permissions.
+        # https://github.com/mitsuhiko/werkzeug/issues/629
```
Comment (wallrj) [inline admin/packaging.py]: This issue was fixed upstream, but hasn't been released yet.

File: admin/packaging.py [hunk 3]
```diff
@@ -478,6 +523,146 @@ def release(self):
...
+            package=self.package, rpm_version=self.rpm_version,
+            architecture=self.architecture)
+
+        output_file = self.destination_path.child(filename)
+
+        try:
+            check_output([
+                {
+                    PackageTypes.RPM: 'rpmlint',
+                    PackageTypes.DEB: 'lintian',
```
Comment (wallrj) [inline admin/packaging.py]: Maybe move this to a global variable.

File: admin/packaging.py [hunk 4]
```diff
@@ -564,6 +749,11 @@ def omnibus_package_builder(
     rpm_version = DelayedRpmVersion(
         package_version_step=get_package_version_step)
 
+    category = {
+        PackageTypes.RPM: 'Applications/System',
+        PackageTypes.DEB: 'admin',
```
Comment (wallrj) [inline admin/packaging.py]: maybe move to a global variable.

**Comment by wallrj** [comment]
There's a build failure:
 * http://build.clusterhq.com/builders/flocker-omnibus-ubuntu-14.04/builds/309/steps/build-package/logs/stdio

```
Traceback (most recent call last):
  File "/flocker/admin/build-package-entrypoint", line 13, in <module>
    docker_main(top_level=TOPLEVEL, base_path=BASEPATH)
  File "/flocker/admin/packaging.py", line 1020, in main
    package_uri=options['package-uri'],
  File "/flocker/admin/packaging.py", line 116, in run
    step.run()
  File "/flocker/admin/packaging.py", line 308, in run
    self._create_virtualenv(root=self.virtualenv.root)
  File "/flocker/admin/packaging.py", line 274, in create_virtualenv
    env=dict(PYTHONDONTWRITEBYTECODE='1')
  File "/flocker/admin/packaging.py", line 131, in run_command
    return check_output(args=args, env=env, cwd=cwd,)
  File "/usr/lib/python2.7/subprocess.py", line 566, in check_output
    process = Popen(stdout=PIPE, *popenargs, **kwargs)
  File "/usr/lib/python2.7/subprocess.py", line 710, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1327, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
program finished with exit code 1
```

I'll try running on my machine and see if I can recreate the failure.

**Comment by wallrj** [comment]
So the Fedora package builder works but the Ubuntu one doesn't. 
It fails because the virtualenv command can't be found.
Which might have to do with its path.

Fedora
```
[root@c6c7fd2ee1b8 /]# type -p virtualenv
/usr/bin/virtualenv

```

vs Ubuntu
```
root@a0c0d4004bd9:/# type -p virtualenv
/usr/local/bin/virtualenv

```

And / or the fact that we call it with ` env=dict(PYTHONDONTWRITEBYTECODE='1'` ....but that wouldn't  affect the PATH for the parent process, so I guess it's not that.

**Comment by wallrj** [comment]
Otherwise, I think this seems fine. 
I verified that the linting is taking place at package build time by commenting out one of the IGNORED lint warnings and saw the command exit with an error code, so buildbot will still pick those up.

```
Package errors (clusterhq-python-flocker):
clusterhq-python-flocker.x86_64: E: zero-length /opt/flocker/lib/python2.7/no-global-site-packages.txt
(1153)(package-lint-FLOC-1153 ✕? =)[~/.../HybridLogic/flocker]$ echo $?
1

```

Please merge after fixing the ubuntu package builder.

**Comment by exarkun** [comment]
> root@a0c0d4004bd9:/# type -p virtualenv
> /usr/local/bin/virtualenv

That seems like a pretty terrible place for it to be.  I wonder why it's there.

**Comment by tomprince** [comment]
> That seems like a pretty terrible place for it to be. I wonder why it's there.

https://bazaar.launchpad.net/~doko/python/pkg2.7-debian/view/head:/patches/distutils-install-layout.diff#L79 I think.

**Pull request merged**